### PR TITLE
Fixing memory corruption and some minor bugs

### PR DESCRIPTION
**Pull request by noskill**

### CODE REVIEW

**Comment by noskill** [comment]
at_bfs_f-ffparser crashes with "double free or corruption (out)" error on many problems from http://icaps-conference.org/ipc2008/deterministic/data/domains/netben-opt-compiled-2.tar.bz2 . For example woodworking-numeric/07

**Comment by miquelramirez** [comment]
Hello @noskill,

thanks for the PR. I have been reviewing your changes yesterday and today I hope I will find time to make some comments.

Regarding the numeric planning problems you mention, we don't have support for them on LAPKT at the moment. Full support for that kind of more expressive planning task is in the works, but I cannot provide an ETA for its publication at the moment, still early days. In the meantime and in parallel I'd suggest you to look on the following:

 - See how hard would be to reuse [TFD](http://gki.informatik.uni-freiburg.de/tools/tfd/)'s front-end, which should allow you to account for numeric state variables and expressions discussed on the specification of [PDDL 2.1](https://www.jair.org/media/1129/live-1129-2132-jair.pdf).

 - Re-implement the concepts and data structures in the agnostic backend of LAPKT so you can have states with discrete and numeric variables. You'd also need to define a new StateModel class.

I hope this helps,

Miquel.

File: external/fd/pddl/conditions.py [hunk 1]
```diff
@@ -277,7 +277,10 @@ class Literal(Condition):
     parts = []
     def __init__(self, predicate, args):
         self.predicate = predicate
-        self.args = tuple(args)
+        if any(isinstance(x, list) for x in args):
+            self.args = tuple(tuple(x) for x in args)
+        else:
+            self.args = tuple(args)
```
Comment (miquelramirez) [inline external/fd/pddl/conditions.py]: Hi @noskill,

how does this relate to the double free issue? Also, what is the rationale for the change above? When can the `args` parameter contain instances of `list`?

File: planners/at_bfs_f-ffparser/src/ipc2014_rwa.hxx [hunk 2]
```diff
@@ -29,26 +29,26 @@ namespace aptk {
...
 	class Node {
 	public:
 		typedef Fibonacci_Open_List< Node >	Open_List;
-		typedef	State				State_Type;	
+		typedef	State				State_Type;
+
+		typedef Node<State>* Node_Ptr;
+		typedef typename std::list< Node<State>* >                     	Node_Ptr_List;
+		typedef typename std::list< Node<State>* >::reverse_iterator    	Node_Ptr_List_Rit;
+		typedef typename std::list< Node<State>* >::iterator            	Node_Ptr_List_It;
```
Comment (miquelramirez) [inline planners/at_bfs_f-ffparser/src/ipc2014_rwa.hxx]: Hello again @noskill,

thanks for this Anatoly. This code was put together in a rather haphazard way to meet the IPC 2014 competition deadline, and it shows.

Miquel.

**Comment by noskill** [comment]
Hi @miquelramirez! 

in h_1.hxx i changed 
v > 0.0f to v >= 0.0f 

This change i am least sure about.  I've written small domain with soft goals. at_bfs_f-ffparser gets stuck in an infinite loop in function compute(line 315) when v=0 in the call to update function.

My domain and problem:
https://gist.github.com/noskill/f162354db7b3c6dc3dafd34898572d6d
https://gist.github.com/noskill/f04f0fcdfb3e03ae71d0ec7e55125c97

**Comment by miquelramirez** [comment]
Hi Anatoly,

we're still quite busy working for ICAPS, as I imagine you may be as well. We'll get back to you after the deadline!

**Comment by noskill** [comment]
Well, maybe one day I will..

**Comment by miquelramirez** [comment]
Hello @noskill,

the issue is related to having actions with zero cost... which is an assumption necessary to have a well-defined notion of "best supporter" (i.e. so you cannot have an action to be the best supporter to one of its own preconditions). The observed infinite loop happens because zero value supporters are taken to "trump" existing best supporters.  The code extracting relaxed plans depends on the best supporters relations to form an acyclic graph in order to terminate.

I would suggest very much that you use a very small "epsilon" cost on actions without costs (like 0.0001) to avoid this boundary condition.

On the other hand, the fix for the RWA* algorithm is perfectly fine.

Miquel.

**Comment by noskill** [comment]
Hi! 
Can we merge it? I might create another request then, currently working on pddl generation...

**Comment by miquelramirez** [comment]
Thanks very much Anton for your contributions and sorry about not getting back to you earlier, I have been on holidays for a while!
